util/network-devp2p/src/host.rs

```diff
@@ -214,10 +215,48 @@ impl<'s> NetworkContextTrait for NetworkContext<'s> {
 	}
 }
 
+struct WrappedEnr {
```

Comment:
I don't like this API quite a lot, for various reasons:
1. We use private fields, since it's defined in the same file.
2. The naming is quite unfortunate. I think we should simply wrap the entire `Enr` library and isolate it's use to a single module where the wrapper is defined. That way all `enr` specific stuff will not leak to `network-devp2p`, and will be nicely contained - this way will also allow you to just call this type `Enr` without getting into conflicts.
3. We've spent a great deal of time getting rid of `Deref` implementations, cause they are super confusing in large codebase. Even rust docs state that `Deref` should be used only in very restricted cases, not for arbitrary wrappers.
4. The `edit()` with `Guard` stuff is really ugly, especially that it leads to a construct like this:
`drop(enr.edit())`

I'd go with:
```
pub struct Enr {
   inner: enr::Enr,
   key: enr::DefaultKey,
}

impl Enr {
  pub fn set_tcp_socket(&mut self) {
    self.save()
  }

  fn save() { 

  }
}
```
so that all mutable operations are just done via `impl` not outside and we can make sure it's saved at the right moment internally.
Also the `enr` library should not leak, but rather we should expose all the things that are necessary via public functions/getters.
And obviously it deserves a file on it's own to avoid polluting already large `host.rs` file.